uc.c

```diff
@@ -935,200 +931,63 @@ MemoryRegion *memory_mapping(struct uc_struct* uc, uint64_t address)
-                                  return UC_ERR_HOOK_EXIST;
-
-                              i = hook_find_new(uc);
-                              if (i) {
-                                  uc->hook_callbacks[i].callback = callback;
-                                  uc->hook_callbacks[i].user_data = user_data;
-                                  *evh = i;
-                                  uc->hook_in_idx = i;
-                                  return UC_ERR_OK;
-                              } else
-                                  return UC_ERR_NOMEM;
-                     case UC_X86_INS_SYSCALL:
-                     case UC_X86_INS_SYSENTER:
-                              // only one event handler at the same time
-                              if (uc->hook_syscall_idx)
-                                  return UC_ERR_HOOK_EXIST;
-
-                              i = hook_find_new(uc);
-                              if (i) {
-                                  uc->hook_callbacks[i].callback = callback;
-                                  uc->hook_callbacks[i].user_data = user_data;
-                                  *evh = i;
-                                  uc->hook_syscall_idx = i;
-                                  return UC_ERR_OK;
-                              } else
-                                  return UC_ERR_NOMEM;
-                 }
-                 break;
-    }
-
-    return UC_ERR_OK;
-}
-
 UNICORN_EXPORT
 uc_err uc_hook_add(uc_engine *uc, uc_hook *hh, int type, void *callback, void *user_data, ...)
 {
     va_list valist;
     int ret = UC_ERR_OK;
-    int id;
-    uint64_t begin, end;
 
     va_start(valist, user_data);
 
-    if (type & UC_HOOK_MEM_READ_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_READ_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_PROT, callback, user_data, hh);
+    struct hook *hook = calloc(1, sizeof(struct hook));
+    if (hook == NULL) {
+        return UC_ERR_NOMEM;
+    }
+    hook->type = type;
+    hook->callback = callback;
+    hook->user_data = user_data;
+    hook->refs = 0;
+    *hh = (uc_hook)hook;
+
+    // everybody but HOOK_INSN gets begin/end, so exit early here.
+    if (type & UC_HOOK_INSN) {
+        hook->insn = va_arg(valist, int);
+        hook->begin = 1;
+        hook->end = 0;
+        if (list_append(&uc->hook[UC_HOOK_INSN_IDX], hook) == NULL) {
+            free(hook);
+            return UC_ERR_NOMEM;
+        }
+        hook->refs++;
```

Comment:
is this "refs" properly implemented? i fail to see where we can bump "refs" up after the first time (?)